`[Experimental]` `[WIP]` Transparent Compiler

0101 opened the pull request:
# Transparent Compiler

## Synopsis

This is an attempt to replace the current `BackgroundCompiler` (powered by [`IncrementalBuilder`](https://github.com/dotnet/fsharp/blob/main/src/Compiler/Service/IncrementalBuild.fs)) with a new system that will be easier to understand and change and that will also be in line with https://github.com/dotnet/fsharp/issues/11976

### Issues with current Background Compiler
What bothers me a lot with current `IncrementalBuilder` / `GraphNode` / `BoundModel` system is the opaqueness of the workflow. It's often difficult to follow the code because it suddenly calls a graph node, which contains a _computation_ which was constructed at some prior point but now there's no link to the actual code that will be executed. Hence the name _Transparent compiler_ and the attempt to make this better 😄 (mostly I had no good idea for what to call this).

### How is this different

The idea is to replace the `GraphNode` with a sort of cache/memoization, so that the computation can be accessed from anywhere given the correct key/input. Instead of having to have a reference to a pre-constructed graph which needs to be stored and updated. 

So the resulting code should look mostly as if we're calculating everything from scratch for every single request - hopefully simple and with no mis-direction - except some intermediate steps (and the whole result as well) will be cached/memoized. 

### Will this work?
🤷 I don't know, but I want to try. 

## Goals

### Functional goals

- Work with editor buffers instead of the filesystem (in a more natural way than the current "live buffers" solution which relies on notifications and a callback function to retrieve file contents)
- Leverage [dependency graph](https://github.com/dotnet/fsharp/blob/main/src/Compiler/Driver/GraphChecking/Docs.md) to avoid unnecessary computation and enable parallelization 

### Non-functional goals

- Immutable compilation model - avoid state in the background compiler. Get everything we need as input and return a result. 
- Easily understandable code

### Possible future benefits

- When we move to out-of-process/LSP model, caches from parsing and type checking in the IDE can be re-used in building the assemblies

## Technical details

### Caching / memoization

We need to deal with concurrency and cancellation and we don't want to repeat the same computations. To do this we use a similar system that was already implemented in `GraphNode`, [but never used](https://github.com/dotnet/fsharp/pull/14908). The cache will store a `Task` (or `node`, still to be decided) and will serialize all requests with ~a `MailboxProcessor`~ an async lock which makes sure to only start the computation once and keep track of request count to see if it should be canceled. 

### API 

We need a new format for parsing/checking requests that will contain everything needed to perform the operations. The major difference to the current API is replacing `FSharpProjectOptions.SourceFiles` file, which is now just a string with path, with a record that will also have a `Version` (which we will fully rely on when caching) and a `GetSource: unit -> Task<ISourceText>` which we will call if we need to get the contents. 

The output/results should remain the same.

### Transition from current background compiler

The new API is a superset of the old one so we can keep all the old code and switch between them with a configuration flag. This way we can offer it to early adopters for testing with a way to switch back if it doesn't work properly. 
`
Also during development `TransparentCompiler` keeps it's own instance of `BackgroundCompiler` and uses it for workflows that are not yet implemented, so it can be tested before completing everything.

## Status

It's a work in progress. Feel free to have a look and give any kind of feedback! 

You can also check out this branch, build it and try it out (in VS, it can be enabled in F# advanced options). It should more or less work for F#-only solutions that don't do anything fancy.

Some of the things that still need to be done:


- [ ] Enable tests with TC in CI
- [ ] More benchmarks
- [ ] Support for in-memory C# project references
- [ ] Support for scripts
- [ ] Support for type providers
- [ ] Preserve stack traces in `AsyncMemoize`
- [ ] Enable requesting 2 levels of checking a file - fastest possible to get diagnostics and semantic classification - and all-inclusive for all IDE features, such as completions that could include things from files we don't depend on
- [x] Plug in to existing tests for `FSharpChecker`
- [x] Figure out merging `TcInfo`s from files processed in parallel
- [x] Support for `itemKeyStore` and other stuff from `TcInfoExtras`
- [x] Reuse `TcIntermediate` results in ParseAndCheckFileInProject if available 
- [x] Make it work properly with signature files with [`NodeToTypeCheck`](https://github.com/dotnet/fsharp/blob/124ca6bd0257ac8c594abe0390befd0d039d7f13/src/Compiler/Driver/ParseAndCheckInputs.fs#L1428-L1456)
- [x] Support for in-memory F# project references
- [x] Support for on-disk project references
- [x] Make sure we don't hold on to stuff that should be GC'd in `AsyncMemoize`
- [x] Implement all background compiler APIs
- [x] Allow using a DLL on disk if it's up to date rather than doing a in-memory type check of that project
- [x] Don't (strongly) keep different versions of the same thing in the caches
- [ ] ~`?` Add some types around `AsyncMemoize` to prevent accidentally using the wrong key vs. computation input~
- [ ] ~`?` Get rid of `NodeCode` and use just `Task` (maybe cancellable task builder?)~ need it for diagnostics scopes
- [ ] ~More efficient implementation of `leafSequence`~ don't need it
- [ ] ~`?` Figure out merging `TcInfo`s from files processed in parallel without callbacks~ too hard

## Some preliminary synthetic benchmarks

``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 10 (10.0.19045.2846)
AMD Ryzen 9 5900X, 1 CPU, 24 logical and 12 physical cores
.NET SDK=7.0.203
  [Host]     : .NET 7.0.5 (7.0.523.17405), X64 RyuJIT AVX2 DEBUG
  Job-EQVOVN : .NET 7.0.5 (7.0.523.17405), X64 RyuJIT AVX2

InvocationCount=1  IterationCount=4  UnrollFactor=1  
WarmupCount=1  

```
|  |     ProjectType |       Mean |       Error |    StdDev | Completed Work Items |        Gen0 |      Gen1 |  Allocated |
 |----------------------- |---------------- |-----------:|------------:|----------:|-----------------:|-----------:|----------:|-----------:|
 |                  **Incremental Builder** | **Dependency Chain** | **3,931.7 ms** |    **97.63 ms** |  **15.11 ms** |             **156** |  **14000** | **2000** | **5471.11 MB** |
 |                  **Incremental Builder** | **Dependent Groups** | **3,882.0 ms** |   **160.07 ms** |  **24.77 ms** |             **150** |  **14000** | **2000** | **5440.74 MB** |
 |                  **Incremental Builder** |  **Parallel Groups** | **4,142.5 ms** |   **259.68 ms** |  **40.19 ms** |             **150** |  **14000** | **2000** | **5624.65 MB** |
 |                   **Transparent Compiler** | **Dependency Chain** | **3,954.3 ms** | **1,549.37 ms** | **239.77 ms** |            **1308** |   **9000** | **2000** | **3466.58 MB** |
 |                   **Transparent Compiler** | **Dependent Groups** |   **560.2 ms** |    **79.75 ms** |  **12.34 ms** |             **867** |   **1000** |         **-** |  **553.02 MB** |
 |                   **Transparent Compiler** |  **Parallel Groups** | **1,318.1 ms** |    **79.63 ms** |  **12.32 ms** |            **1164** |   **9000** | **3000** | **3518.59 MB** |

majocha commented on src/Compiler/Service/TransparentCompiler.fs:
```diff
@@ -0,0 +1,987 @@
+                DeduplicateParsedInputModuleName prevTcInfo.moduleNamesDict input
+
+            let! (tcEnvAtEndOfFile, topAttribs, implFile, ccuSigForFile), tcState =
+                CheckOneInput(
```
Probably `CheckOneInputWithCallback` will be the way to go. With some care to not get duplicated stuff in prior tcState.

0101 replied:
I would prefer to solve it without callbacks for increased simplicity. But if it doesn't work out then I will probably use that one.

majocha commented:
I don't have anything clever to say but this looks awesome and absolutely should work!

majocha commented on src/Compiler/Service/TransparentCompiler.fs:
```diff
@@ -0,0 +1,987 @@
+
+            use _ = new CompilationGlobalsScope(diagnosticsLogger, BuildPhase.Parameter)
+
+            // TODO: might need to put something like this somewhere
```
Yes all inputs should be somehow versioned and contribute to project snapshot identity.

0101 replied:
This is now handled by `ReferencedOnDisk` in project snapshot.

majocha commented on src/Compiler/Service/TransparentCompiler.fs:
```diff
@@ -670,11 +673,27 @@ type internal TransparentCompiler
+            let implFile = implFileOpt.Value
+
+            let _ccuSigForFile, tcState =
+                AddCheckResultsToTcState
```
If two files share a dependency upstream, do we need to avoid adding it twice to the state?

0101 replied:
Possibly. This is still highly experimental 😄 

Though now the simplest tests are passing without having to deal with that.

0101 replied:
This is now handled by the callback system.

0101 commented:
I added some basic synthetic project benchmarks (results in the description). Looks promising so far. In the _DependencyChain_ project where each file depends on the previous one the performance is similar to IncrementalBuilder, but with less allocations and memory consumed. In the other scenarios where we can leverage graph checking by either parallelization or skipping unneeded work we can see the expected time savings.

There is significantly more **Completed Work Items**. If it has no other side effects then it probably won't bother us but might need to look into that...

TheAngryByrd commented:
@0101 let me know when you think this API is pretty stable and I'll give it a run in FSAC.

vzarytovskii commented on src/Compiler/Driver/GraphChecking/Graph.fs:
```diff
@@ -27,26 +27,42 @@ module internal Graph =
-        let transitiveDeps (node: 'Node) =
-            let visited = HashSet<'Node>()
+    let nodes (graph: Graph<'Node>) : Set<'Node> =
+        graph.Values |> Seq.collect id |> Seq.append graph.Keys |> Seq.distinct |> Set
```
It doesn't need call to distinct, does it? Set should take care of it, as long as there's equality.

vzarytovskii commented on src/Compiler/Facilities/AsyncMemoize.fs:
```diff
@@ -0,0 +1,122 @@
+    let cache =
+        MruCache<_, 'TKey, Job<'TValue>>(keepStrongly = 10, areSame = (fun (x, y) -> x = y))
+
+    let requestCounts = Dictionary<'TKey, int>()
```
Shouldn't this be a concurrent dictionary?

0101 commented on src/Compiler/Driver/GraphChecking/Graph.fs:
```diff
@@ -59,6 +75,48 @@ module internal Graph =
         |> readOnlyDict
         |> addIfMissing originalGraph.Keys
 
+    /// Returns leaves of the graph and the remaining graph without the leaves.
```
Should be ok unless there's a project with _a lot_ of files. But it can probably be optimized together with `leafSequence`, it's already in the open tasks.

0101 commented:
### Status update

After the latest tweaks it seems to work with [Giraffe](https://github.com/giraffe-fsharp/Giraffe) solution. Which is the simplest one I found that contains some real world useful code :) So if someone has some smaller projects like that, you could already try it out. 

It doesn't quite work with VisualFSharp solution yet, but I'll be now looking at what are the missing pieces for that...

auduchinok commented:
> It doesn't quite work with VisualFSharp solution yet, but I'll be now looking at what are the missing pieces for that...

There's some special logic around having FSharp.Core as a project in a solution. You may want to try using it on FSharp.Compiler.Service.sln in a fresh working copy, it doesn't contain that project reference.

0101 commented:
> @0101 Would it be possible to run the benchmarks on it too? 🙂

Yeah I'm hoping to do that, but not sure exactly how yet. So far my idea was to enable the SyntheticProject / workflow builder to load real projects from disk and then work with them in a similar way.

0101 commented:
Managed to load Giraffe into a benchmark. Thanks to @TheAngryByrd for [SDK reference loading trick](https://github.com/TheAngryByrd/IcedTasks/blob/master/generate-sdk-references.fsx)!

I tried to benchmark some random file modifications and checking. The performance is better than IncrementalBuilder, but not dramatically. 

Giraffe project doesn't actually allow for that much dependency graph optimization. 

<details>
	<summary>Giraffe Depndency Graph</summary>


```mermaid

flowchart RL

    0["FileHelpers.fs"]

    1["FileShortGuid.fs"]

    2["FileDateTimeExtensions.fs"]

    4["FileJson.fs"]

    5["FileXml.fs"]

    6["FileFormatExpressions.fs"]

    7["FileModelParser.fs"]

    8["FileHttpContextExtensions.fs"]

    9["FileCore.fs"]

    10["FileResponseCaching.fs"]

    11["FileModelValidation.fs"]

    12["FileAuth.fs"]

    13["FileRouting.fs"]

    14["FilePreconditional.fs"]

    15["FileStreaming.fs"]

    16["FileNegotiation.fs"]

    17["FileHttpStatusCodeHandlers.fs"]

    18["FileMiddleware.fs"]

    19["FileEndpointRouting.fs"]

    1 --> 0

    2 --> 0

    4 --> 0

    4 --> 2

    5 --> 0

    5 --> 2

    6 --> 0

    6 --> 1

    6 --> 2

    7 --> 0

    7 --> 2

    8 --> 0

    8 --> 2

    8 --> 4

    8 --> 5

    8 --> 7

    9 --> 0

    9 --> 2

    9 --> 8

    10 --> 0

    10 --> 2

    10 --> 8

    10 --> 9

    11 --> 0

    11 --> 2

    11 --> 8

    11 --> 9

    11 --> 10

    12 --> 0

    12 --> 2

    12 --> 8

    12 --> 9

    12 --> 10

    12 --> 11

    13 --> 0

    13 --> 2

    13 --> 6

    13 --> 7

    13 --> 8

    13 --> 9

    13 --> 10

    13 --> 11

    13 --> 12

    14 --> 0

    14 --> 2

    14 --> 8

    14 --> 9

    14 --> 10

    14 --> 11

    14 --> 12

    14 --> 13

    15 --> 0

    15 --> 2

    15 --> 8

    15 --> 9

    15 --> 10

    15 --> 11

    15 --> 12

    15 --> 13

    15 --> 14

    16 --> 0

    16 --> 2

    16 --> 8

    16 --> 9

    16 --> 10

    16 --> 11

    16 --> 12

    16 --> 13

    16 --> 14

    16 --> 15

    17 --> 0

    17 --> 2

    17 --> 8

    17 --> 9

    17 --> 10

    17 --> 11

    17 --> 12

    17 --> 13

    17 --> 14

    17 --> 15

    17 --> 16

    18 --> 0

    18 --> 2

    18 --> 4

    18 --> 5

    18 --> 8

    18 --> 9

    18 --> 10

    18 --> 11

    18 --> 12

    18 --> 13

    18 --> 14

    18 --> 15

    18 --> 16

    18 --> 17

    19 --> 0

    19 --> 1

    19 --> 2

    19 --> 8

    19 --> 9

    19 --> 10

    19 --> 11

    19 --> 12

    19 --> 13

    19 --> 14

    19 --> 15

    19 --> 16

    19 --> 17

    19 --> 18

```

</details>



But it's still a bit faster and consumes less memory. And there's still room for more optimizations, so far I've been mostly trying to get it to work at all 😄. 

[Benchmark code](https://github.com/0101/fsharp/blob/6f703c82ae824f2a48792cba347420d98e87f9fa/tests/benchmarks/FCSBenchmarks/CompilerServiceBenchmarks/BackgroundCompilerBenchmarks.fs#L374-L428)

|                Method | Use Transparent Compiler |       Mean |     Error |    StdDev | Completed Work Items | Lock Contentions |      Gen0 |      Gen1 |  Allocated |
|---------------------- |----------------------- |-----------:|----------:|----------:|---------------------:|-----------------:|----------:|----------:|-----------:|
|  **Change First Check Last** |                  **False** |   **675.0 ms** |  **88.11 ms** |  **46.08 ms** |              **40** |                **-** | **2000** | **1000** |  **471.07 MB** |
| Change Second Check Last |                  False |   650.1 ms |  83.40 ms |  43.62 ms |              39 |                - | 2000 | 1000 |  464.28 MB |
|          Some Workflow |                  False | 1,660.6 ms |  97.36 ms |  43.23 ms |              68 |                - | 5000 | 2000 | 1160.21 MB |
|  **Change First Check Last** |                   **True** |   **645.3 ms** | **231.55 ms** | **121.10 ms** |             **203** |                **-** | **2000** | **1000** |  **430.01 MB** |
| Change Second Check Last |                   True |   358.2 ms |  30.83 ms |  16.12 ms |             187 |                - | 1000 |         - |  219.71 MB |
|          Some Workflow |                   True | 1,384.2 ms |  36.42 ms |  19.05 ms |            1428 |                - | 4000 | 1000 |  955.96 MB |


<details>
  <summary>Potato Laptop specs</summary>

``` ini
BenchmarkDotNet=v0.13.2, OS=Windows 11 (10.0.22621.1848)
11th Gen Intel Core i7-11850H 2.50GHz, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.203
  [Host]     : .NET 7.0.7 (7.0.723.27404), X64 RyuJIT AVX2 DEBUG
  Job-QMBAKL : .NET 7.0.7 (7.0.723.27404), X64 RyuJIT AVX2

IterationCount=8  WarmupCount=1  
```

</details>

0101 commented:
@nojaf sure, I'll add support for signature files and try it out. Although the benchmark might not make sense as is with the signature files.

0101 commented:
@nojaf Here are the results with signature files. 

Feel free to propose some other workflows to benchmarks.

|       Method | Use Transparent Compiler | Signature Files |       Mean |     Error |   StdDev | Completed Work Items |      Gen0 |      Gen1 |  Allocated |
|------------- |----------------------- |--------------- |-----------:|----------:|---------:|---------------------:|----------:|----------:|-----------:|
| **SomeWorkflow** |                  **False** |          **False** | **1,622.3 ms** |  **99.48 ms** | **52.03 ms** |              **68.0000** | **5000** | **2000** | **1162.12 MB** |
| **SomeWorkflow** |                  **False** |           **True** |   **925.4 ms** | **111.80 ms** | **58.47 ms** |              **80** | **2000** | **1000** |  **575.92 MB** |
| **SomeWorkflow** |                   **True** |          **False** | **1,329.0 ms** |  **26.23 ms** | **11.65 ms** |            **1439** | **4000** | **1000** |  **958.08 MB** |
| **SomeWorkflow** |                   **True** |           **True** |   **667.0 ms** |  **97.70 ms** | **43.38 ms** |            **2182** | **2000** | **1000** |  **385.51 MB** |

nojaf commented:
Thank you @0101, these are very interesting numbers!

safesparrow commented on src/Compiler/Driver/GraphChecking/Graph.fs:
```diff
@@ -27,26 +27,42 @@ module internal Graph =
+        dfs node
+        visited |> Seq.toArray
 
+    let transitive<'Node when 'Node: equality> (graph: Graph<'Node>) : Graph<'Node> =
```
Somewhat related:

Both old (sorry) and new code do lots of unnecessary work for `transitive`.
Each node should be visited exactly once, whereas we visit them multiple times, because:
1. we start from scratch for each node. Instead, we should traverse the graph once, and build results for each node based on results of its dependencies. Would reduce time complexity from ~O(n*n) to O(n).
2. In `transitiveDeps`, we only check `visited` before we run `dfs` for all dependencies - instead, a simple loop with checking the `visited` membership each time would mean much fewer calls

safesparrow commented on src/Compiler/Service/TransparentCompiler.fs:
```diff
@@ -0,0 +1,1647 @@
+                    ComputeDependencyGraphForFile priorSnapshot (parsedInputs |> Array.map p13) bootstrapInfo.TcConfig
+
+                // layers that can be processed in parallel
+                let layers = Graph.leafSequence graph |> Seq.toList
```
Processing nodes in layers might make scheduling simpler, but it does reduce parallelisation compared with processing each node as soon as all its dependencies have finished.

Was this a conscious decision? It's fine for an MVP, but with potential for improvement 🙂

0101 replied:
I just wanted to make it as simple as possible and to get it working. But in the end if we can make it faster we should. Might be nice to reuse the scheduling you made here if it's possible. 

But I want to make it work properly first before doing much optimization.

0101 replied:
Actually, it's not so simple anyway. I was having some issues with duplication from ArtificialImplFiles. So I tried to plug in your graph processing and it seems to work fine. I duplicated some of it for now because of NodeCode/Async, but maybe we'll be able to unify it.

safesparrow commented on src/Compiler/Driver/GraphChecking/Graph.fs:
```diff
@@ -27,26 +27,42 @@ module internal Graph =
+    /// Get subgraph of the given graph that contains only nodes that are reachable from the given node.
+    let subGraphFor node graph =
+        let allDeps = graph |> transitiveDeps node
+        let relevant n = n = node || allDeps |> Array.contains n
```
This is O(n*n) - we could optimise by returning a `HashSet` from `transitiveDeps` and using it here.

0101 replied:
Yes, I just did this quickly and moved on. On sufficiently large projects it might make a difference so we should definitely optimize it.

0101 commented:
# Status Update

I'm still working on this!

At the moment it's kinda usable for VisualFSharp solution (I'm using it every day), though there are still unresolved issues.

## Debugging

I've been having some difficulties reproducing issues I've been seeing in simple projects (and outside of VS), so I've invested some time in making it more debug friendly in regular use on our big solution - via attaching another VS instance. Most of FCS work is contained in the `AsyncMemozie` instances, which now provide some statistics, allow inspecting running jobs and collect exceptions (of which there should be eventually 0).

<img width="950" alt="TCCacheDebugging" src="https://github.com/dotnet/fsharp/assets/217092/51b10cea-c4a1-4543-9b88-c42db7ef2863">


## Caching system update

- I Added new LRU Cache supporting versions so we can allow GC of document and project related entries once new versions are observed without having to wait for cache capacity to be depleted.
- It's based on Dictionary and doubly linked lists so it should be pretty efficient when storing a lot of entries (like parse and typecheck results for all files in the solution)
- Also added hashing of cache keys (versions) - they could get pretty big with a lot of project references

### Concurrency and async

- I experimented with a bunch of ways of serializing the cache/memoize requests
  - Got rid of `MailboxProcessor` and replaced it with a Task-based one and tried to use `CancellableTasks`, but ended up with weird issues of double-setting of results or exceptions which seemed originate in CancellableTask state machine
  - Tried using just a regular lock, because the amount of work which needs to be synchronized is very small and should be superfast, but still ended up with deadlocks in VS
  - At the moment there's an "async lock" made from SemaphoreSlim which seems to work ok
- What I think we _might_ need but hopefully not
  - Some kind of prioritization of tasks, so that cache requests - which are processed very quickly always (you get back an async/task which either is already done/running or we just started it), don't get delayed when the thread pool is fully busy running `Cancellable` type-checking tasks which don't yield

### Stacktraces

I've tried to preserve stack traces but didn't really figure that out. Though I think at this level it's not so critical, didn't really miss them much when debugging, so setting this aside for now. Would be good to solve it eventually if possible.

## Other improvements

- Added support for on-disk references (tracking DLL modified times), good option for big solutions if DLLs are already built
  - Eventually we should be also able to tell whether a DLL on disk is up to date with the sources we have and then use it, otherwise do the typechecking in memory
- Switched Graph processing (fork) to use `TaskCompletionSource` instead of cancellation to signal end of processing and exceptions
  - I've had some issues where a "completion" cancellation propagated back into the individual graph processing jobs and caused some havoc

## VS gotchas

### Document versions

One weird thing VS does is report a different source version for a `Document` once it's been _open_ in the editor. Even though the contents didn't change. This was causing a lot of unnecessary work upon just opening files even though we already parsed and checked them as a part of previous work.

I mitigated this by adding `ProjectSnapshotWithSources` as a key for parsing and subsequent operations, using SourceText's checksum to tell if they changed.

But this points out a whole area of the design that needs to be better thought out - because we should be doing the same with any subsequent results, like syntax tree and typed tree, to not trigger recomputation when nothing _meaningful_ has changed.

### Generating project snapshots

Initially I just slapped in a quick generation of a `ProjectSnapshot` from `Document` and `FSharpProjectOptions` which we already have a system for. Turned out it wasn't fast enough and really lagged out VS when triggering _Find all references_ or a _Code search_. Optimizing this really improved the usability. But there are still questions about how the IDEs should be handling this and what the final shape of `ProjectSnapshot` should be.


## The APIs

Some thinking is still needed about what the final public APIs will look like. Whether it will be more like now, a request/response or input/output operations with FCS caching intermediate results and figuring out what can be reused from before. Or having a big opaque project snapshot containing all the intermediate results (`CompilerCaches`) and leaving it up to the user to keep it synchronized with files in the IDE and on disk.

Some of the considerations here are:

- How it's going to work with LSP
- Will it enable us to have a proper _background_ compiler which can reuse results also between IDE operations and builds
  - So when you e.g. edit a file and want to run a test, not everything has to be built from scratch


## The annoying errors

There are still errors which need to be tracked down.

### Doubled types

These are happening _I suspect_ when some results get kicked out of cache and we have to mix some cached results with some newly computed ones. It looks like we get some doubled types:

```
Error	FS0001	This expression was expected to have type     'String'     but here has type     'string'
```

or

```
Error	FS0193

The type
'Microsoft.FSharp.Core.int (FSharp.Core, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)'
does not match the type
'Microsoft.FSharp.Core.int (FSharp.Core, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)'
```

Hopefully this isn't some intrinsic issue with Graph checking and caching individual results.


### Race condition issues

There are still some bugs in `AsyncMemoize` or `LRUCache` which pop out from time to time under some unknown conditions which I couldn't replicate in isolation. Like dictionary modified when writing into it (even though it's under a lock) or Running jobs missing in cache even though we explicitly check for that before removing anything.

These should be easier to fix and in theory might resolve also the above, but I somehow doubt that.


## Trying this out

### In VS

Anyone brave enough can build a VSIX from this branch and install it in VS to try it out. It still contains our present and fully functional `IncrementalBuilder` and you can switch between the two in the options (though it requires VS restart). 

Easy mode is with Enable in-memory cross project references turned off.

I also added a Cache Size Factor input in the options which says how much will be strongly held in the caches. Hopefully this setting won't be needed in the end. But for now it might be useful if you're either running out of memory or solution is too big and things are being recomputed when they shouldn't.

- Type providers will most likely not work, didn't look at that yet
- Also scripts might not work
- And also any other features which we don't have in VisualFSharp solution or I didn't come across them yet

### Out of VS

Since the APIs will most likely change, potentially a lot, I wouldn't recommend spending too much time writing any code to deal with just FCS at this point.

majocha commented on src/Compiler/Service/FSharpCheckerResults.fsi:
```diff
@@ -41,6 +45,183 @@ type DelayedILModuleReader =
+        /// Timestamp of project/script load, used to differentiate between different instances of a project load.
+        /// This ensures that a complete reload of the project or script type checking
+        /// context occurs on project or script unload/reload.
+        LoadTime: DateTime
```
In CI / test scenarios the resolution of system clock can be too low, leading to some heisenbugs:
https://stackoverflow.com/questions/2143140/c-sharp-datetime-now-precision
I can't recall where was it exactly, but I remember running into this problem at some point in the IncrementalBuild refactor.

0101 replied:
Good job I'm not really using this then 😄. 

But thanks for the heads up, I'll make sure not to rely on that anywhere.

0101 commented:
@dotnet/fsharp-team-msft I think we can merge this for an early preview and continue improving it with separate PRs. Please have a look, mainly focusing on whether anything can impair our current setup in main.

psfinaki commented on azure-pipelines.yml:
```diff
@@ -439,6 +439,10 @@ stages:
               vs_release:
                 _configuration: Release
                 _testKind: testVs
+              # experimental_features:
```
Honestly I am not a fan adding more commented code to our pipelines.

0101 replied:
Yeah I can revert this. It's more for reference, but not the best solution anyway.

psfinaki commented:
@0101 please post the latest and greatest benchmark results before merging - practice shows it will be helpful for future reference.

nojaf commented:
Thanks for sharing these numbers @0101! Great work!

vzarytovskii commented on azure-pipelines.yml:
```diff
@@ -86,11 +86,11 @@ stages:
         parameters:
           MirrorRepo: fsharp
-          MirrorBranch: release/dev17.9
+          MirrorBranch: release/dev17.8
```
that's not right, should be 17.9

vzarytovskii commented on azure-pipelines.yml:
```diff
@@ -86,11 +86,11 @@ stages:
   #-------------------------------------------------------------------------------------------------------------------#
   - ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
-    - ${{ if eq(variables['Build.SourceBranch'], 'refs/heads/release/dev17.9') }}:
+    - ${{ if eq(variables['Build.SourceBranch'], 'refs/heads/release/dev17.8') }}:
```
Also should be 17.9

0101 replied:
I accidentally reverted this file to an outdated `main` I had locally. But I already fixed it, there's no changes to the azure-pipelines in the PR anymore.

vzarytovskii commented on azure-pipelines.yml:
```diff
@@ -818,8 +806,8 @@ stages:
-      componentBranchName: refs/heads/release/dev17.9
-      insertTargetBranch: rel/d17.9
+      componentBranchName: refs/heads/release/dev17.8
+      insertTargetBranch: main
```
also should be 17.9

Smaug123 commented on tests/FSharp.Compiler.ComponentTests/CompilerService/AsyncMemoize.fs:
```diff
@@ -101,6 +101,8 @@ let ``We can cancel a job`` () =
 
         Assert.Equal<(JobEvent * int) array>([| Started, key |], eventLog |> Seq.toArray )
 
+        do! Task.Delay 100
```
Is there a way to phrase this with explicit gates within the computations so that we know when execution has progressed into the tasks, and then the tasks block on the cancellation of the CTS? Unless there's some pretty mad shenanigans somewhere, these tests are inherently flaky (#TimeIsNotASynchronisationPrimitive). I would expect at some point the failwith on line 80 to be hit, if nothing else, when during some test run the task scheduler happens not to schedule the main thread (which performs the cancellation) for one second; such coincidences are inevitable when running large test suites many times.

0101 replied:
Yeah, you're right. I should figure out how to do it without arbitrary time delays. Probably use the events somehow to progress the test.

Smaug123 replied:
(There's nothing wrong with a helper function that spins waiting for a sequence to contain n elements, for example, although since one must tediously ensure the collection is safe to read and write concurrently, for tests I usually wrap it up in a type that controls the accesses made to it and which has a `Freeze : unit -> 'a list` method.)

vzarytovskii commented:
@0101 Shall this be merged, or is there any pending changes?

0101 commented:
@vzarytovskii I want to improve the tests as pointed out by Smaug123, but it can also be done separately.

vzarytovskii commented:
> @vzarytovskii I want to improve the tests as pointed out by Smaug123, but it can also be done separately.

Up to you, but if you ask me, we should merge now to not end up in a never ending cycle of fixing small things and never merging.

auduchinok commented:
> Yeah, you're right. I should figure out how to do it without arbitrary time delays. Probably use the events somehow to progress the test.

I've tried to update our FCS fork and also found some instability, seemingly related to requests cancelation. My theory is it may be #16348 to blame, but I haven't an opportunity to look into it further yet. It seems that the previous approach from https://github.com/dotnet/fsharp/pull/16137 works as expected (though, some access to metadata is not guarded with cancellation tokens).

@0101 Could you try to check if reverting #16348 would improve stability of the tests?

vzarytovskii commented:
@auduchinok fwiw (not really related to this PR per se), once this is merged, I will be rewriting NodeBuilder to state machines (pretty much cancellable task), so cancellation will change a bit there.

0101 commented:
> > Yeah, you're right. I should figure out how to do it without arbitrary time delays. Probably use the events somehow to progress the test.
> 
> I've tried to update our FCS fork and also found some instability, seemingly related to requests cancelation. My theory is it may be #16348 to blame, but I haven't an opportunity to look into it further yet. It seems that the previous approach from #16137 works as expected (though, some access to metadata is not guarded with cancellation tokens).
> 
> @0101 Could you try to check if reverting #16348 would improve stability of the tests?

I wouldn't expect that to be a problem, since the tests don't run in parallel. I just didn't spend much effort on writing the tests properly, so I still want to do that first.

Smaug123 commented:
> Up to you, but if you ask me, we should merge now to not end up in a never ending cycle of fixing small things and never merging.

I have a counter-opinion, but I commit very rarely to this repo so downweight it accordingly.

I consider flaky tests to be *seriously* bad (like "drop everything and fix this; if nothing else just delete the test" bad). Their effect is to very rapidly cause people to distrust all the tests, they greatly increase the friction of contributing, and they essentially prevent most automation on the repo; and the reputation of a project "having flaky tests" is extremely hard to reverse. To my surprise, Dan Luu has not actually written directly about this, but he has written on [https://danluu.com/broken-builds/](broken builds) and [the normalisation of deviance](https://danluu.com/wat/).

majocha commented on src/Compiler/Facilities/AsyncMemoize.fs:
```diff
@@ -0,0 +1,611 @@
+
+    let processStateUpdate post (key: KeyData<_, _>, action: StateUpdate<_>) =
+        task {
+            do! Task.Delay 0
```
`Task.Delay` with 0 ms is a special case and just returns `Task.CompletedTask`.

`Task.Yield ()`?

0101 replied:
Whoops, yeah, that's what I wanted. Wonder if it's actually necessary.

auduchinok commented on src/Compiler/Service/TransparentCompiler.fs:
```diff
@@ -0,0 +1,2105 @@
+#endif
+                        return tcImports
+                    with
+                    | :? OperationCanceledException ->
+                        // if it's been canceled then it shouldn't be needed anymore
+                        return frameworkTcImports
```
@0101 Are you sure it's safe to return this value here instead of rethrowing the exception? Can this value get cached somewhere?
